File: solutions/ocaml/binary-search/1/binary_search.ml
let rec f arr target left right = 
    if left >= right then 
        if arr.(left) = target then Ok left else Error "value not in array"
     else
        let middle = (left + right + 1) / 2 in
        if arr.(middle) = target then Ok middle else
            if arr.(middle) < target 
                then f arr target middle right
    else f arr target left (middle - 1)


let find arr target =
    if Array.length arr = 0 then Error "value not in array" else 
    f arr target 0 ((Array.length arr) - 1)

File: solutions/c/binary-search/1/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length)
{
    if (length == 0)
    {
        return NULL;
    }
    size_t l = 0;
    size_t r = length - 1;
    while (l < r)
    {
        size_t middle = (l + r + 1) / 2;
        if (arr[middle] <= value)
        {
            l = middle;
        }
        else
        {
            r = middle - 1;
        }
    }
    if (value == arr[l])
    {
        return &arr[l];
    }
    else
    {
        return NULL;
    }
}